block-layer/codecs/encrypted.md

```diff
@@ -0,0 +1,83 @@
+# Specification: Encrypted Block Codec
+
+**Status: Prescriptive - Draft**
+
+This document describes codecs for IPLD blocks (CID + Data) that are encrypted. The
+multicodec idenfier for the cipher and the initital vector are included in the block
+format and parsed into a standardized data model representation.
+
+The following known ciphers may be referenced in encrypted blocks:
+
+| name | multicodec |
+| --- | --- |
+| aes-gcm | 0x1401 |
+| aes-cbc | 0x1402 |
+| aes-ctr | 0x1403 |
+
+## What this spec is not
+
+This is not a complete system for application privacy. The following issues are
+out of scope for this specification, although they can obviously leverage these codecs:
+
+* Key signaling
+* Access controls
+* Dual-layer encryption w/ replication keys
+
+How you determine what key to apply to an encrypted block will need to be done in the
+application layer. How you decide to encrypt a graph and potentially link the encrypted
+blocks together for replication is done at the application layer. How you manage and access
+keys is done in the application layer.
+
+## Encode/Decode vs Encrypt/Decrypt
+
+The goal of specifying codecs that are used for encryption is to allow the codecs to
+include encryption and decryption programs alongside the codec. However, encrypting and
+decrypting are done by the user and are not done automatically as part of any encode/decode
+operation in the codec.
+
+The encryption program returns a data model value suitable for the block encode program. The
+decode program provides a data model value suitable for the decryption program. And the decryption
+program provides a data model value suitable for parsing into a new block (CID and Bytes). These
+programs are designed to interoperate but it's up to the user to combine them and provide the
+necessary key during encryption and decryption.
+
+## Encrypted Block Format
+
+An encrypted block can be decoded into its initializing vector and the encrypted byte
+payload. Since the initializing vector is a known size for each AES variant the block
+format is simply the iv followed by the byte payload.
+
+```
+| varint(cipher-multicodec) | varint(ivLength) | iv | bytes |
+```
+
+This is decoded into IPLD data model of the following schema.
+
+```ipldsch
+type AES struct {
+  code Int
```

Comment:
Perhaps I would've chosen to try experimenting with `@type` instead of `code` to see how it feels since it might encourage library development that takes nominative typing into account (e.g. #71 uses `@type: codecNumber` as an example).

However, I definitely understand just wanting to roll something out for encryption without worrying about longer term ramifications or bikeshedding on what `@type` could look like.

Reply:
`code` has been how we typically refer to multicodecs in a variety of locations. in this case, these all correspond to multicodecs so it makes sense.

this representation ends up being kind of important in `js-multiformats` because it consistently destructures to the correct property name across the system

Reply:
This seems very potäto/potato to me.  I sincerely doubt making a field called "@type" instead of "code" here is going to influence anyone's hypothetical future library development in any clear direction.

Also: "@type" would not be syntactically valid IPLD Schema syntax: symbols are not permitted in field names.  And for this, the reason in turn is: suppose someone wants to feed this into codegenerators: that "@" isn't going to be a valid field name in _most_ programming languages under the sun, either.

Reply:
I'm likely missing lots of context on `js-multiformats` but

> because it consistently destructures to the correct property name across the system

is the point I was getting at with `@type`, the more structures adopt a convention like that this the easier it becomes to reuse it in the future. (e.g. if `@type` defined could be defined as a union of { Int | String | CID ... } similar to the proposal ).

>Also: `@type` would not be syntactically valid IPLD Schema syntax: symbols are not permitted in field names. And for this, the reason in turn is: suppose someone wants to feed this into codegenerators: that "@" isn't going to be a valid field name in most programming languages under the sun, either.

That's interesting and seems to push even more towards building libraries where a field like `@type` is considered special since it's a valid data model field name but not a valid schema name, and we've been pushing people towards describing their data with schemas.

Reply:
I’ll try to clarify.

In `js-multiformats` there are tools for defining and working with codecs, hashers, base encoders and *now* encryption/decryption programs.

In all cases the “lookup” for a given implementation is done with an integer code. Our implementations of CID and multihash include a `code` property that matches a `code` property on codec and hash implementations. Looking up a given implementation is easy and consistent since you can do `const lookup = ({ code }) => { /* return implementation */ }` and have it work consistently across all multiformats.

That’s why it’s so useful to have the data model representation use a `code` property because we can then lookup the decryption program in multiformats by just passing the return value from `decode()` directly to a lookup function that already works with CIDs and multihashes.

We aligned all of these in the last major refactor when we moved to integer codes in order to avoid needing to ship every codec and base encoder, and the multiformats table, in order to support string names for codecs and hashers.